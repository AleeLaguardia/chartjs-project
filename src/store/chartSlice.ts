import { createSlice, PayloadAction } from '@reduxjs/toolkit';

interface ChartState {
  data: number[]
};

const initialState: ChartState = {
  data: [10, 20, 30, 40, 50],
};

const chartSlice = createSlice({
  name: 'chart',
  initialState,
  reducers: {
    setData: (state, action: PayloadAction<number[]>) => {
      state.data = action.payload;
    },
  },
});

export const { setData } = chartSlice.actions;

export default chartSlice.reducer;