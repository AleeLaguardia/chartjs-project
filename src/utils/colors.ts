export const white = '#FFFFFF';
export const black = '#000000';
export const wildWatermelon = '#FF6384';
export const pictionBlue = '#36A2EB';
export const goldenTainoi = '#FFCE56';
export const fountainBlue = '#4BC0C0';
export const heliotrope = '#9966FF';
export const yellowOrange = '#FF9F40';