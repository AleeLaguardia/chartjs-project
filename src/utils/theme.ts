import { DefaultTheme } from "styled-components";
import {
  white,
  black,
  fountainBlue,
  goldenTainoi,
  heliotrope,
  pictionBlue,
  wildWatermelon,
  yellowOrange
} from './colors';

export const theme: DefaultTheme = {
  colors: {
    white,
    black,
    fountainBlue,
    goldenTainoi,
    heliotrope,
    pictionBlue,
    wildWatermelon,
    yellowOrange
  },
  fonts: {},
};