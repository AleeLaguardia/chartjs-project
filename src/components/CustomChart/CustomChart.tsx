import React from 'react';
import { Chart } from 'react-chartjs-2';
import { Chart as ChartJS, CategoryScale, LinearScale, BarElement, Title, Tooltip, Legend } from 'chart.js';
import annotationPlugin from 'chartjs-plugin-annotation';
import zoomPlugin from 'chartjs-plugin-zoom';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../../store';
import { setData } from '../../store/chartSlice';
import { theme } from '../../utils/theme';

ChartJS.register(CategoryScale, LinearScale, BarElement, Title, Tooltip, Legend, annotationPlugin, zoomPlugin);

const CustomChart: React.FC = () => {
  const dispatch = useDispatch();
  const chartData = useSelector((state: RootState) => state.chart.data);

  const { wildWatermelon, pictionBlue, goldenTainoi, fountainBlue, heliotrope, yellowOrange } = theme.colors;

  const data = {
    labels: ['Red', 'Blue', 'Yellow', 'Green', 'Purple', 'Orange'],
    datasets: [
      {
        label: '# of Votes',
        data: chartData,
        backgroundColor: [
          `${wildWatermelon}50`,
          `${pictionBlue}50`,
          `${goldenTainoi}50`,
          `${fountainBlue}50`,
          `${heliotrope}50`,
          `${yellowOrange}50`,
        ],
        borderColor: [
          wildWatermelon,
          pictionBlue,
          goldenTainoi,
          fountainBlue,
          heliotrope,
          yellowOrange,
        ],
        borderWidth: 1,
      },
    ],
  };

  const options = {
    responsive: true,
    plugins: {
      legend: {
        position: 'top' as const,
      },
      title: {
        display: true,
        text: 'Custom Chart',
      },
      annotation: {
        annotations: {
          line1: {
            type: 'line' as const,
            yMin: 20,
            yMax: 40,
            borderColor: wildWatermelon,
            borderWidth: 2,
          },
        },
      },
      zoom: {
        zoom: {
          wheel: {
            enabled: true,
          },
          pinch: {
            enabled: true,
          },
          mode: 'x' as const,
        },
        pan: {
          enabled: true,
          mode: 'x' as const,
        },
      },
    },
    onHover: (event: any, elements: any) => {
      if (elements.length) {
        const { index } = elements[0];
        console.log(`Hovered over ${data.labels[index]}`);
      }
    },
    onClick: (event: any, elements: any) => {
      if (elements.length) {
        const { index } = elements[0];
        console.log(`Clicked on ${data.labels[index]}`);
        dispatch(setData(chartData.map((item, position) => (position === index ? item + 1 : item))));
      }
    },
  };

  return <Chart type='bar' data={data} options={options} />;
};

export default CustomChart;
