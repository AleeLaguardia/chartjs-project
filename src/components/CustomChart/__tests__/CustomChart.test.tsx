import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import chartSlice from '../../../store/chartSlice';
import CustomChart from '../CustomChart';
import { ThemeProvider } from 'styled-components';
import { theme } from '../../../utils/theme';

const store = configureStore({
  reducer: {
    chart: chartSlice,
  },
});

jest.mock('hammerjs');

describe('Unit test for CustomChart Component', () => {
  it('Renders with default props and styles', () => {
    const { asFragment } = render(
      <ThemeProvider theme={theme}>
        <Provider store={store}>
          <CustomChart />
        </Provider>
      </ThemeProvider>
    );
    
    expect(asFragment).toMatchSnapshot();
  });
});