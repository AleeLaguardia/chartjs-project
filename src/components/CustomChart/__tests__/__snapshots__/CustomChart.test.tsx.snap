// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for CustomChart Component Renders with default props and styles 1`] = `[Function]`;
